src/crypto/crypto_hash.cc

```diff
@@ -34,22 +35,86 @@ void Hash::MemoryInfo(MemoryTracker* tracker) const {
+  // real/original algorithm name.
+  // We use EVP_*_fetch() as a filter here because it will only return an
+  // instance if the algorithm is supported by the public OpenSSL APIs (some
+  // algorithms are used internally by OpenSSL and are also passed to this
+  // callback).
+  EVP_MD* explicit_md = EVP_MD_fetch(nullptr, real_name, nullptr);
+  if (!explicit_md) return;
+#endif  // OPENSSL_VERSION_MAJOR >= 3
+
+  Environment* env = static_cast<Environment*>(arg);
+  env->supported_hash_algorithms.push_back(from);
+
+#if OPENSSL_VERSION_MAJOR >= 3
+  env->evp_md_cache.emplace_back(explicit_md);
+#endif  // OPENSSL_VERSION_MAJOR >= 3
+}
+
+const std::vector<std::string>& GetSupportedHashAlgorithms(Environment* env) {
+  if (env->supported_hash_algorithms.empty()) {
+    MarkPopErrorOnReturn mark_pop_error_on_return;
+    std::vector<std::string> results;
+    EVP_MD_do_all_sorted(CacheSupportedHashAlgorithms, env);
+#if OPENSSL_VERSION_MAJOR >= 3
+    CHECK_EQ(env->supported_hash_algorithms.size(), env->evp_md_cache.size());
+    CHECK_GT(env->supported_hash_algorithms.size(), 0);
+#endif  // OPENSSL_VERSION_MAJOR >= 3
+  }
+  return env->supported_hash_algorithms;
+}
+
 void Hash::GetHashes(const FunctionCallbackInfo<Value>& args) {
-  Environment* env = Environment::GetCurrent(args);
-  MarkPopErrorOnReturn mark_pop_error_on_return;
-  CipherPushContext ctx(env);
-  EVP_MD_do_all_sorted(
+  Local<Context> context = args.GetIsolate()->GetCurrentContext();
+  Environment* env = Environment::GetCurrent(context);
+  const std::vector<std::string>& results = GetSupportedHashAlgorithms(env);
+
+  Local<Value> ret;
+  if (ToV8Value(context, results).ToLocal(&ret)) {
+    args.GetReturnValue().Set(ret);
+  }
+}
+
+const EVP_MD* GetDigestImplementation(Environment* env,
+                                      Local<Value> algorithm,
+                                      Local<Value> algorithm_id) {
+  CHECK(algorithm->IsString());
+  CHECK(algorithm_id->IsInt32());
+  int32_t id = algorithm_id.As<Int32>()->Value();
+
+  const std::vector<std::string>& algorithms = GetSupportedHashAlgorithms(env);
+  std::string algorithm_str;
+
+  if (id != -1) {
 #if OPENSSL_VERSION_MAJOR >= 3
-    array_push_back<EVP_MD,
-                    EVP_MD_fetch,
-                    EVP_MD_free,
-                    EVP_get_digestbyname,
-                    EVP_MD_get0_name>,
+    CHECK_LT(static_cast<size_t>(id), algorithms.size());
+    auto& ptr = env->evp_md_cache[id];
+    CHECK_NOT_NULL(ptr.get());
+    return ptr.get();
 #else
-    array_push_back<EVP_MD>,
-#endif
-    &ctx);
-  args.GetReturnValue().Set(ctx.ToJSArray());
+    algorithm_str = algorithms[id];
+#endif  // OPENSSL_VERSION_MAJOR >= 3
+  }
+
+  if (algorithm_str.empty()) {  // It could be unsupported algorithms.
+    Utf8Value utf8(env->isolate(), algorithm);
+    algorithm_str = utf8.ToString();
+  }
+  const EVP_MD* implicit_md = EVP_get_digestbyname(algorithm_str.c_str());
```

Comment:
Another thing that needs to be taken care of is, when we are dealing with certain aliases (e.g. sha-1 instead of SHA1) we should try `EVP_MD_fetch `using `EVP_MD_get0_name(implicit_md)` and update the cache/mapping again. Apparently we can't get all the aliases in `EVP_MD_do_all_sorted()` and have to a bit of on-demand caching of aliases ourselves or otherwise there would be a performance cliff when users use aliases.

Reply:
Actually I noticed that `sha-1` is not a supported alias so currently on the main branch, this would fail:

```js
require('crypto').createHash('sha-1');  // Throws an error
```

But this would not

```js
require('crypto').createHash('sha1');  // Add all digest names at the same time
require('crypto').createHash('sha-1');  // No longer throws an error because alias can be found
```

And after this patch, the first one would not throw anymore since we eagerly load all the names via `getHashes()` upon loading of the crypto module.